🎨 Improved What's New to hide updates for new users

rob-ghost opened the pull request:
## Why this change?

When new users first access Ghost admin, they see notifications for changelog entries published before they created their account. This creates confusion because they're being notified about updates to a baseline they never experienced - everything is new to them.

Ref https://linear.app/ghost/issue/DES-1155

## What does this change do?

Automatically sets a default `lastSeenDate` for new users on their first visit to the admin app. This ensures they only receive notifications for updates published after they start using Ghost, rather than seeing notifications for every historical changelog entry.

**Technical implementation:**

When the `whatsNew` service loads for the first time and detects that a user doesn't have a `lastSeenDate` in their accessibility settings, it now automatically sets it to the current date (in UTC ISO format) and persists this to the user model. This happens during the initial `fetchLatest` call using a read-merge-write pattern to preserve other accessibility settings.

## Why is this needed?

New users shouldn't be overwhelmed with notifications about changes they never experienced. This provides eventual consistency for the What's New feature across all users while ensuring notifications remain relevant and contextual.

## Implementation approach

This PR follows a test-driven development approach with three commits:

1. **Test coverage** - Added unit and acceptance tests for the What's New feature, establishing a test baseline for the service
2. **Refactor** - Simplified the service to track only the `whatsNew` slice of user accessibility settings rather than the entire object
3. **Feature** - Implemented automatic default `lastSeenDate` initialization for new users with tests

rob-ghost committed:
Added test coverage for What's New feature

ref https://linear.app/ghost/issue/DES-1155

- added unit tests for whats-new service
- added acceptance tests for toast, modal, and badge behavior
- added data-test attributes to templates for reliable test selectors

rob-ghost committed:
Refactored whats-new service to own only whatsNew settings slice

ref https://linear.app/ghost/issue/DES-1155

- removed whatsNewSettings computed property (was redundant passthrough)
- changed service to track only whatsNew slice instead of entire user.accessibility
- updated updateLastSeen to use read-merge-write pattern for safe persistence

rob-ghost committed:
🎨 Improved What's New to hide updates for new users

ref https://linear.app/ghost/issue/DES-1155

New users will now only see What's New notifications for updates published after their first visit to the admin app, preventing confusion about historical updates they haven't experienced.

The system automatically sets lastSeenDate to the current date on first load, ensuring eventual consistency across all user accounts.

rob-ghost commented:
👋🏻 @kevinansfield @jonatansberg

Here's the 3rd attempt at this feature (as proposed [here](https://github.com/TryGhost/Ghost/pull/25041#issuecomment-3362111605)).

I think out of the three options ([first the FE depending on onboarding state](https://github.com/TryGhost/Ghost/pull/25014), [second defaulting in the backend](https://github.com/TryGhost/Ghost/pull/25041)) I think this approach is my preference.

It's got low coupling between `whats-new` and other components / services / layers, and high cohesion within the service. It keeps all the `whats-new` concerns co-located in one service.

The drawback compared to the backend approach is the upcoming Ember -> React migration, and adding some more code to be ported, however it will be much easier to port given everything is located in one place.

---

The first two commits add both unit tests for the `whats-new` service, and acceptance tests for the various ways of viewing or interacting with the "What's new" notification and modals (including adding `data-test-*` attributes to the templates), as well as a refactor of the service to simplify its state management.

The third commit would be easiest to review, since its the commit which adds the defaulting of the `lastSeenDate` to the `whatsNew` accessibility state.

rob-ghost commented on ghost/admin/app/components/modals/whats-new.hbs:
```diff
@@ -1,17 +1,17 @@
-<div class="modal-content">
-    <h1 class="gh-whasnew-modal-title">What’s new?</h1>
```
Great spot! You have eagle eyes 👀

Claude ran away with itself when it couldn't get the tests to reliably pass with the apostrophes and it replaced them and I missed it.

I ran into this problem on the other PR too. The solution is to update the test to handle it and not update the template. I'll fix this.

kevinansfield commented on ghost/admin/tests/acceptance/whats-new-test.js:
```diff
@@ -0,0 +1,358 @@
+            mockChangelog(this.server, [ENTRIES.newFeatured]);
+
+            await visit('/site');
+            await settled();
```
We shouldn't need settled() in almost all cases, the ember test helpers call this internally before resolving so we can clean these extraneous uses up 🙂 If we find they are needed to make a test pass it suggests we have a deeper issue somewhere.

rob-ghost commented on ghost/admin/app/components/modals/whats-new.hbs:
```diff
@@ -1,17 +1,17 @@
-<div class="modal-content">
-    <h1 class="gh-whasnew-modal-title">What’s new?</h1>
-    <section class="gh-whatsnew-modal-entries" {{did-insert (perform this.whatsNew.updateLastSeen)}}>
+<div class="modal-content" data-test-whats-new-modal>
```
Much prefer this myself 👍🏻 changed the `data-test` attributes and references in the page objects.

jonatansberg reviewed (commented):
I don't have much to add in addition to Kevin's review of the fix. The approach looks solid! 👍 

I have some meta commentary on adding 800 lines of new tests for whats supposedly a minor bugfix. That makes me wonder if we might have gone a little overboard with the testing?

What I mean by that is, does all of the tests add value and make this code easier to maintain in the future, to port to React eventually? We're testing behaviors we care about? Do some of these tests add a future maintenance burden without telling us anything that either the unit tests or the acceptance tests would catch anyhow? 

Take the test on error handling for example. To me, the behavior I would like to verify is that we fail gracefully if something blows up. Exactly if and how we log to the console is rather inconsequential. 

Similarly, tests for the `_whatsNewSettings` state maybe aren't all that desirable? The underscore naming suggests we're testing what was at some point meant to be an implementation detail. 

Make of these observations what you will. The meta meta discussion probably ought to be around whether or not it's worthwhile to invest more time into this particular thing at all :)

jonatansberg commented:
It's all good. I'm sorry if my previous comment came off as a bit grumpy 😅 

It's definetly valuable to add tests since there hasn't been any. It also makes sense to invest time in getting familiar with things, and I think we should continue to make room for that as we go 😊

rob-ghost commented:
@jonatansberg I did a pass on the tests and change the acceptance tests to be pure behaviour of the toast and modal, so it doesn't know anything about the fact that its driven by `lastSeenDate` or anything else, just that the toast appears (or doesn't), and disappears (even after refreshes).

I also removed the suite for the internal `_whatsNewSettings` and adjusted the graceful error handling tests to be purely about error handling.

I toyed with removing the badge acceptance tests but I think they're interesting to keep.

I'm happy that the tests now are just testing outcomes and they're not tightly coupled with the implementation of the service.

rob-ghost committed:
Refactored What's New initialization to centralize defaults (#25100)

ref https://linear.app/ghost/issue/DES-1155

Moved default lastSeenDate initialization from fetchLatest to `init()`,
making the service state predictable from creation and simplifying the
initialization logic.

This improves maintainability by ensuring the service is always in a
valid state, eliminates defensive fallback code, and makes the
separation of concerns clearer: `init()` handles setup, `fetchLatest()`
handles data loading/syncing, and `updateLastSeen()` handles user
interactions.
